### PR TITLE
HPCC-21301 Fix some test differences seen on larger clusters.

### PR DESCRIPTION
**Pull request by jakesmith**
Larger cluster configurations exposed issues with a couple of
the regression tests, that produced inconsistent results,
depending on the size of the cluster (e.g. due to ordering issues).

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by jakesmith** [comment]
@ghalliday - please review

File: testing/regress/ecl/prefetch2.ecl [hunk 1]
```diff
@@ -40,12 +40,12 @@ END;
 
 outrec trans1(in L) := TRANSFORM
   SELF.word := L.word;
-  SELF.doc := SORTED(Files.getSearchSuperIndex()(KEYED(kind = TS.kindType.TextEntry),keyed(word = l.word)), doc)[1].doc;
+  SELF.doc := SORT(Files.getSearchSuperIndex()(KEYED(kind = TS.kindType.TextEntry),keyed(word = l.word)), doc)[1].doc;
```
Comment (ghalliday) [inline testing/regress/ecl/prefetch2.ecl]: This is reading a result from a superkey of two indexes.  The original result is showing a match from the first of the subfiles in the superkey.  I suspect Thor was hitting a match from the second index which occurred on an earlier node.
So
i) the original SORTED() is wrong
ii) Is it well defined which order matches come in from subindexes of a superindex?

@richardkchapman what is your opinion?

Comment (richardkchapman) [reply testing/regress/ecl/prefetch2.ecl]: I think subindexes are supposed to interleave - i.e. you should see matches from part 1 of subindex 1 then from part 1 of subindex 2 then from part 2 of subindex 1

Comment (jakesmith) [reply testing/regress/ecl/prefetch2.ecl]: > I suspect Thor was hitting a match from the second index which occurred on an earlier node.

Confirmed that is what's happening.
Thor sees a match in the 2nd index 1st (on an earlier node).

Comment (jakesmith) [reply testing/regress/ecl/prefetch2.ecl]: So Thor is interleaving. In this case the width the Thor is the same as the indexes.
So the 1st slave, is processing the 1st parts of the index.
The 2nd slave is processing the 2nd parts, etc.

Thor sees a match from the 2nd index first, the 1st index has matches, but they come in partitions read on later nodes, so are not part of the i[1].doc result.

Which I think is correct.

Comment (jakesmith) [reply testing/regress/ecl/prefetch2.ecl]: Btw, hthor result is same as roxie. So it looks like both read all parts of subfile1, before reading subfile2 ..

**Comment by richardkchapman** [comment]
What happens if you add a SORTED,ASSERT?

If there's a SORTED on there but the data is not, in fact, sorted, then the result is undefined, I would have thought

**Comment by jakesmith** [comment]
SORTED,ASSERT doesn't have any effect.
But that's because it is sorted to the point it has read up to, because of the SELECTNTH(1).

If you change the query to try to output all matches, so it reads whole index with SORTED,ASSERT,
e.g.:
```
SORTED(Files.getSearchSuperIndex()(KEYED(kind = TS.kindType.TextEntry),keyed(word = 'boy')), doc, ASSERT);
```

then all engines fail (as they should) with 'detected incorrectly sorted rows'

**Comment by richardkchapman** [comment]
typo in commit message
@ghalliday

**Comment by ghalliday** [comment]
@jakesmith ok, seems reasonable to merge it.  Can you add a #onwarning to suppress the difference in roxie.

**Comment by jakesmith** [comment]
@ghalliday - please merge.

**Pull request merged**